driver/src/contracts/base_contract.rs

```diff
@@ -22,26 +34,81 @@ impl BaseContract {
         let web3 = web3::Web3::new(transport);
 
         let json: serde_json::Value = serde_json::from_str(&contents)?;
-        let abi: String = json.get("abi").ok_or("No ABI for contract")?.to_string();
+        let abi_string = json.get("abi").ok_or("No ABI for contract")?.to_string();
+        let abi = abi_string.as_bytes();
 
         let decoded_address = hex::decode(&address[2..])?;
         let contract_address: Address = Address::from(&decoded_address[..]);
-        let contract = Contract::from_json(web3.eth(), contract_address, abi.as_bytes())?;
+        let contract = Contract::from_json(web3.eth(), contract_address, abi)?;
+
+        let network_id = env::var("NETWORK_ID")?.parse()?;
+        let private_key = H256::from_str(&env::var("PRIVATE_KEY")?)?;
 
         Ok(BaseContract {
             contract,
             web3,
             event_loop,
+            abi: ethabi::Contract::load(abi)?,
+            network_id,
+            public_key: address_from_private_key(&private_key)?,
+            private_key,
         })
     }
 
-    pub fn account_with_sufficient_balance(&self) -> Option<Address> {
-        let accounts: Vec<Address> = self.web3.eth().accounts().wait().ok()?;
-        accounts
-            .into_iter()
-            .find(|&acc| match self.web3.eth().balance(acc, None).wait() {
-                Ok(balance) => !balance.is_zero(),
-                Err(_) => false,
-            })
+    pub fn send_signed_transaction<P>(
+        &self,
+        function: &str,
+        params: P,
+        options: Options,
+    ) -> Result<H256>
+    where
+        P: Tokenize,
+    {
+        let nonce = if let Some(nonce) = options.nonce {
```

Comment:
I am not sure I see what the `Some` is doing here. I read it like "If nonce is an attribute of options"... is this what it means?

Reply:
if options.nonce is Some (as in it's not None) then the local variable nonce will have the value of the option (unwrapped).

Reply:
Yes... but I find it weird that you are using `Some(nonce)` where `nonce` doesn't appear to exist yet...

Reply:
It's a different variable (I could call it foo) and basically says whatever the value inside the option is (which is of type Some(value)) should be bound to that variable.

Reply:
Its part of rust's pattern matching:
```rust
if let /* pattern */ {
    // pattern matched, you can use extracted bindings from the pattern here
} else {
    // pattern didn't match
}
```
 the full pattern syntax is here: https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html